subprojects/docs/src/docs/userguide/lazyConfiguration.adoc

```diff
@@ -15,40 +15,226 @@
+
+++++
+<sample id="workingWithFiles" dir="providers/fileAndDirectoryProperty" title="Using file and directory property">
+    <sourcefile file="build.gradle"/>
+    <output args="print"/>
+</sample>
+++++
+
+This example shows how `Provider` types can be used inside an extension. Lazy values for api:org.gradle.api.Project#getBuildDir()[] and api:org.gradle.api.Project#getProjectDir()[] can be accessed through api:org.gradle.api.Project#getLayout()[] with api:org.gradle.api.file.ProjectLayout#getBuildDirectory()[] and api:org.gradle.api.file.ProjectLayout#getProjectDirectory()[].
 
 [[sec:working_with_task_dependencies_in_lazy_properties]]
-=== Working with task dependencies
+=== Working with task dependencies and Providers
+
+Many builds have several tasks that depend on each other. This usually means that one task processes the outputs of another task as an input. For these outputs and inputs, we need to know their locations on the file system and appropriately configure each task to know where to look. This can be cumbersome if any of these values are configurable by a user or configured by multiple plugins.
+
+To make this easier, Gradle offers convenient APIs for defining files or directories as task inputs and outputs in a descriptive way. As an example consider the following plugin with a producer and consumer task, which are wired together via inputs and outputs:
+
+++++
+<sample id="implicitTaskDependency" dir="providers/implicitTaskDependency" title="Implicit task dependency">
+    <sourcefile file="build.gradle"/>
+    <output args="consumer"/>
+</sample>
+++++
+
+In the example above, the task output and input are connected before any location is defined. This is possible because the input and output properties use the `Provider` API. The output property is created with api:org.gradle.api.DefaultTask#newOutputFile()[] and the input property is created with api:org.gradle.api.DefaultTask#newInputFile()[]. Values are only resolved when they are needed during execution. The setters can be called at any time before the task is executed and the change will automatically affect all related input and output properties.
+
+Another thing to note is the absence of any explicit task dependency. Properties created via `newOutputFile()` and `newOutputDirectory()` bring knowledge about which task is generating them, so using them as task input will implicitly link tasks together.
+
+[[sec:working_with_collection]]
+=== Working with collection Providers
+
+In this section, we are going to explore lazy collections. They work exactly like any other `Provider` and, just like `FileSystemLocation` providers, they have additional modeling around them. They are implemented as a strongly typed model called api:org.gradle.api.provider.ListProperty[]. You can create a new `ListProperty` using api:org.gradle.api.model.ObjectFactory#listProperty(java.lang.Class)[] and specifying the element's type.
```

Comment:
Is this basically the equivalent of `Collection<Provider>`?

Reply:
Sort of.  It's more like a `Provider<List<T>>`.  The value is a `List<T>`, we don't have a list of `Provider<T>`.  You can't change the value of a particular position in the list, you can only replace it or append to it.